1963 - support multiple direct_generators

sergiuionescu opened the pull request:
See https://github.com/ruflin/Elastica/issues/1963

deguif commented on src/Suggest/Phrase.php:
```diff
@@ -140,7 +140,7 @@ public function setSmoothingModel(string $model, array $params): Phrase
      */
     public function addCandidateGenerator(AbstractCandidateGenerator $generator): Phrase
```
What about deprecating this method, and instead add a new one `addDirectGenerator(DirectGenerator $generator)` as this suggester only allow configuring `direct_generator`. That would this way avoid to trick to `toArray` method.

ruflin reviewed (commented):
Thanks for the contribution. Could you also add a test and a changelog entry?

Also see comment from @deguif

sergiuionescu commented:
> Thanks for the contribution. Could you also add a test and a changelog entry?
> 
> Also see comment from @deguif

@deguif I added the new method and deprecated the old one since it was a bit confusing to use. The toArray trick was required by `\Elastica\Suggest\CandidateGenerator\DirectGenerator::toArray` which is also updated.

@ruflin I updated the changelog. When do you think you could include this in a patch version?

Thanks.

deguif commented on src/Suggest/Phrase.php:
```diff
@@ -138,30 +139,18 @@ public function setSmoothingModel(string $model, array $params): Phrase
 
     /**
-     * {@inheritdoc}
+     * @deprecated Please use addDirectGenerator instead.
```
```suggestion
     * @deprecated since version 7.2.0, use the "addDirectGenerator()" method instead.
```

deguif commented on src/Suggest/Phrase.php:
```diff
@@ -138,30 +139,18 @@ public function setSmoothingModel(string $model, array $params): Phrase
-        }
-
-        return $array;
+        return $this->addParam('candidate_generator', $generator);
```
```suggestion
        \trigger_deprecation('ruflin/elastica', '7.2.0', 'The "%s()" method is deprecated, use the "addDirectGenerator()" method instead. It will be removed in 8.0.', __METHOD__);

        return $this->addParam('candidate_generator', $generator);
```

deguif commented on src/Suggest/Phrase.php:
```diff
@@ -138,30 +139,18 @@ public function setSmoothingModel(string $model, array $params): Phrase
      * @return $this
      */
-    public function addCandidateGenerator(AbstractCandidateGenerator $generator): Phrase
+    public function addDirectGenerator(DirectGenerator $generator): Phrase
```
```suggestion
    public function addDirectGenerator(DirectGenerator $generator): self
```

deguif commented on src/Suggest/CandidateGenerator/DirectGenerator.php:
```diff
@@ -132,4 +132,20 @@ public function setPostFilter(string $post)
+     *
+     * @return array
+     */
+    public function toArray()
+    {
+        $data = $this->getParams();
+
+        if (!empty($this->_rawParams)) {
+            $data = \array_merge($data, $this->_rawParams);
+        }
+
+        return $this->_convertArrayable($data);
```
Implementing NameableInterface here should do the trick and avoid customizing the `toArray()`

sergiuionescu replied:
@deguif Thanks for suggesting the changes and align this with the future changes for the lib.

 I do not see how implementing NameableInterface will do the trick here. The handling for that interface is:  
`$arr[$value instanceof NameableInterface ? $value->getName() : $key] = $value->toArray();` So we will still get a map with the "direct_generator" name and params, instead of a list of direct_generator's params.

We are looking to get 
```
{
    "suggest": {
        "text": "obel prize",
        "simple_phrase": {
            "phrase": {
                "field": "title.trigram",
                "size": 1,
                "direct_generator": [
                    {
                        "field": "title.trigram",
                        "suggest_mode": "always"
                    },
                    {
                        "field": "title.reverse",
                        "suggest_mode": "always",
                        "pre_filter": "reverse",
                        "post_filter": "reverse"
                    }
                ]
            }
        }
    }
}
```

I think the override is the more transparent approach. Am i missing something with the NameableInterface, do you have an example where that is used to build a list?

sergiuionescu replied:
@deguif Please run the tests whenever possible, i reverted to the `\Elastica\Suggest\Phrase::toArray` override.

deguif replied:
Yes done again, I did it this morning but Github was encountering issues.

deguif reviewed (commented):
Thanks @sergiuionescu for your work.
Some new comments related to how we handle deprecations.
Another one related to `toArray` change that can probably be done another way more standard.

deguif commented on CHANGELOG.md:
```diff
@@ -19,6 +19,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+## [7.1.2](https://github.com/ruflin/Elastica/compare/7.1.1...7.1.2)
+### Added
+* Added `\Elastica\Suggest\Phrase::addDirectGenerator` to align with ES specification [#1964](https://github.com/ruflin/Elastica/pull/1964)
```
This should be moved to the upper `Added section`

deguif commented on CHANGELOG.md:
```diff
@@ -19,6 +19,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 ### Security
 
+## [7.1.2](https://github.com/ruflin/Elastica/compare/7.1.1...7.1.2)
```
This should be removed as it's done upper. Proper version will be edited when preparing the next release.

deguif commented on tests/Suggest/PhraseTest.php:
```diff
@@ -46,6 +46,29 @@ public function testToArray(): void
+
+    /**
+     * @group functional
+     */
```
```suggestion
     * @group legacy
     */
```

deguif replied:
And probably add a check for deprecation triggering (using the proper trait): 
```
$this->expectDeprecation('Since ruflin/elastica 7.2.0: XXXX . It will be removed in 8.0.');
```

deguif commented:
Thanks @sergiuionescu

sergiuionescu commented:
> Thanks @sergiuionescu

Thank you @deguif , @ruflin for the quick action and support!
